File: holoken-app/src/main/java/com/holokenmod/ui/main/GameTopFragment.java
package com.holokenmod.ui.main;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;

import com.holokenmod.R;
import com.holokenmod.creation.GridDifficultyCalculator;
import com.holokenmod.databinding.GameTopFragmentBinding;
import com.holokenmod.game.Game;
import com.holokenmod.options.GameDifficulty;
import com.holokenmod.ui.GridCreationListener;

public class GameTopFragment extends Fragment implements GridCreationListener {

    private Game game;
    
    private GameTopFragmentBinding binding;
    
    private boolean showtimer;
    
    public GameTopFragment() {
        super(R.layout.game_top_fragment);
    }
    
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup parent, Bundle savedInstanceState) {
        binding = GameTopFragmentBinding.inflate(inflater, parent, false);
        
        return binding.getRoot();
    }
    
    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
    
    @Override
    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        if (game != null) {
            freshGridWasCreated();
        }
    }
    
    public void freshGridWasCreated() {
        if (!isAdded()) {
            return;
        }
        
        requireActivity().runOnUiThread( () -> {
            GridDifficultyCalculator difficultyCalculator = new GridDifficultyCalculator(this.game.getGrid());
    
            binding.difficulty.setText(difficultyCalculator.getInfo());
            
            setStarsByDifficulty(difficultyCalculator);
            
            if(this.showtimer) {
                binding.playtime.setVisibility(View.VISIBLE);
            } else {
                binding.playtime.setVisibility(View.INVISIBLE);
            }
        });
    }
    
    public void setGame(Game game) {
        this.game = game;
        
        freshGridWasCreated();
    }
    
    private void setStarsByDifficulty(GridDifficultyCalculator difficultyCalculator) {
        setStarByDifficulty(binding.ratingStarOne,
                difficultyCalculator.getDifficulty(),
                GameDifficulty.EASY);
        setStarByDifficulty(binding.ratingStarTwo,
                difficultyCalculator.getDifficulty(),
                GameDifficulty.MEDIUM);
        setStarByDifficulty(binding.ratingStarThree,
                difficultyCalculator.getDifficulty(),
                GameDifficulty.HARD);
        setStarByDifficulty(binding.ratingStarFour,
                difficultyCalculator.getDifficulty(),
                GameDifficulty.EXTREME);
    }
    
    private void setStarByDifficulty(ImageView view, GameDifficulty difficulty, GameDifficulty minimumDifficulty) {
        if (difficulty.compareTo(minimumDifficulty) >= 0) {
            view.setImageResource(R.drawable.filled_star_20);
        } else {
            view.setImageResource(R.drawable.outline_star_20);
        }
    }
    
    public void setGameTime(String timeDescription) {
        binding.playtime.setText(timeDescription);
    }
    
    public void setTimerVisible(boolean showtimer) {
        this.showtimer = showtimer;
    }
}

File: holoken-app/src/main/kotlin/com/holokenmod/ui/main/GameTopFragment.kt
package com.holokenmod.ui.main

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.fragment.app.Fragment
import com.holokenmod.R
import com.holokenmod.creation.GridDifficultyCalculator
import com.holokenmod.databinding.GameTopFragmentBinding
import com.holokenmod.game.Game
import com.holokenmod.game.GridCreationListener
import com.holokenmod.options.GameDifficulty
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject

class GameTopFragment : Fragment(R.layout.game_top_fragment), GridCreationListener, KoinComponent {
    private val game: Game by inject()
    private lateinit var binding: GameTopFragmentBinding

    private var timeDescription: String? = null
    private var showtimer = false

    override fun onCreateView(
        inflater: LayoutInflater,
        parent: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = GameTopFragmentBinding.inflate(inflater, parent, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        game.addGridCreationListener(this)

        freshGridWasCreated()
    }

    override fun freshGridWasCreated() {
        if (!isAdded) {
            return
        }
        requireActivity().runOnUiThread {
            val difficultyCalculator = GridDifficultyCalculator(game.grid)
            binding.difficulty.text = difficultyCalculator.info()
            setStarsByDifficulty(difficultyCalculator)
            if (showtimer) {
                binding.playtime.visibility = View.VISIBLE
            } else {
                binding.playtime.visibility = View.INVISIBLE
            }

            timeDescription?.let { binding.playtime.text = it }
        }
    }

    private fun setStarsByDifficulty(difficultyCalculator: GridDifficultyCalculator) {
        setStarByDifficulty(
            binding.ratingStarOne,
            difficultyCalculator.difficulty,
            GameDifficulty.EASY
        )
        setStarByDifficulty(
            binding.ratingStarTwo,
            difficultyCalculator.difficulty,
            GameDifficulty.MEDIUM
        )
        setStarByDifficulty(
            binding.ratingStarThree,
            difficultyCalculator.difficulty,
            GameDifficulty.HARD
        )
        setStarByDifficulty(
            binding.ratingStarFour,
            difficultyCalculator.difficulty,
            GameDifficulty.EXTREME
        )
    }

    private fun setStarByDifficulty(
        view: ImageView,
        difficulty: GameDifficulty,
        minimumDifficulty: GameDifficulty
    ) {
        if (difficulty >= minimumDifficulty) {
            view.setImageResource(R.drawable.filled_star_20)
        } else {
            view.setImageResource(R.drawable.outline_star_20)
        }
    }

    fun setGameTime(timeDescription: String) {
        if (this::binding.isInitialized) {
            binding.playtime.text = timeDescription
        } else {
            this.timeDescription = timeDescription
        }
    }

    fun setTimerVisible(showtimer: Boolean) {
        this.showtimer = showtimer
    }
}